### ISSUE TITLE
Doesn't work if component has custom `componentDidUpdate`

### ISSUE THREAD

**Issue by OliverJAsh** [maicki/why-did-you-update#17]
See https://github.com/garbles/why-did-you-update/issues/19 for more detail.

**Comment by danburzo** [comment]
I've just started toying with a different way of monkey-patching it via `React.createElement`:

```js
const getDisplayName = ctor => ctor.displayName || ctor.name || 'Unnamed Component';

const wdyu = React => {
	let oldFn = React.createElement;
	React.createElement = function() {
		let el = oldFn.apply(React, arguments);
		let ctor = el.type;

		if (ctor.prototype && typeof ctor.prototype.render === 'function') {
			// class component
			class WDYUClassComponent extends ctor {
				componentDidUpdate(prevProps, prevState) {
					if (typeof ctor.componentDidUpdate === 'function') {
						ctor.componentDidUpdate(prevProps, prevState);
					}
					// DO MAGIC
				}
			};
			WDYUClassComponent.displayName = getDisplayName(ctor);
			WDYUClassComponent.defaultProps = ctor.defaultProps;
			return {
				...el,
				type: WDYUClassComponent
			}
		} else if (typeof ctor === 'function') {
			// functional component
			class WDYUFunctionalComponent extends React.Component {
				render() {
					return ctor(this.props);
				}

				componentDidUpdate(prevProps, prevState) {
					// DO MAGIC
				}
			}

			WDYUFunctionalComponent.displayName = getDisplayName(ctor);
			WDYUFunctionalComponent.defaultProps = ctor.defaultProps;

			return {
				...el,
				type: WDYUFunctionalComponent
			}
		}
		return el;
	};
}

export default wdyu;
```

this should address the problem and offer a way to observe functional components as well (https://github.com/maicki/why-did-you-update/issues/10)

I still need to work out if the classes I create (WDYUClassComponent and WDYUFunctionalComponents) should be memoized based on the original constructor — not sure yet if it causes problems with reconciliation.

**Comment by danburzo** [comment]
Initial version here: https://github.com/maicki/why-did-you-update/compare/master...danburzo:add-hoc

It's passing most tests, will look into the failing ones. Also need to add back the React.createClass patch.

Should we test it against multiple versions of React as part of the test suite? Seems to be possible to have [multiple versions of a package with Yarn](https://github.com/npm/npm/issues/5499#issuecomment-337968750).

**Comment by danburzo** [comment]
I've the final version of this idea on this pull request: https://github.com/maicki/why-did-you-update/pull/20

**Comment by vzaidman** [comment]
fixed by #20
